### PR TITLE
[SCB-701] RequestBody(required = false) should invoke method when bod…

### PR DESCRIPTION
**Pull request by acsukesh**
…y is null

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java [hunk 1]
```diff
@@ -77,7 +83,10 @@ public void setValue(RestClientRequest clientRequest, Object arg) throws Excepti
       try (BufferOutputStream output = new BufferOutputStream()) {
         clientRequest.putHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
         RestObjectMapper.INSTANCE.writeValue(output, arg);
-        clientRequest.write(output.getBuffer());
+        if (arg != null) {
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java]: Can you add a test case to show return body is required while users pass a null value? Json will encoding null to "null", so I think here may got some problem.

File: common/common-rest/src/test/java/org/apache/servicecomb/common/rest/codec/param/TestBodyProcessor.java [hunk 2]
```diff
@@ -61,11 +61,11 @@
   Buffer outputBodyBuffer;
 
   private void createProcessor(Class<?> type) {
-    processor = new BodyProcessor(TypeFactory.defaultInstance().constructType(type));
+    processor = new BodyProcessor(TypeFactory.defaultInstance().constructType(type), false);
```
Comment (liubao68) [inline common/common-rest/src/test/java/org/apache/servicecomb/common/rest/codec/param/TestBodyProcessor.java]: We must not change a test case if it is required. And here should equivalent to true

File: demo/demo-schema/src/main/resources/microservices/pojo/server.yaml [hunk 3]
```diff
@@ -150,7 +150,7 @@ paths:
       parameters:
         - name: TestRequest
           in: body
-          required: true
+          required: false
```
Comment (liubao68) [inline demo/demo-schema/src/main/resources/microservices/pojo/server.yaml]: We must not change a test case if it is required.  Please add a new one if possible.

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java [hunk 4]
```diff
@@ -221,17 +222,33 @@ private static void testConfigurationDuplicate() {
...
+    user.setName("world");
+    TestMgr.check("ha world",
+        template.postForObject(prefix + "/annotations/saysomething?prefix={prefix}",
+            user,
+            String.class,
+            "ha"));
+
+    TestMgr.check("No user data found",
+        template.postForObject(prefix + "/annotations/saysomething?prefix={prefix}",
+            null,
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java]: As mentioned in above comments, "null" may not indicates an empty body. Need to distinguish null or empty.

File: swagger/swagger-generator/generator-springmvc/src/test/resources/schemas/mixupAnnotations.yaml [hunk 5]
```diff
@@ -39,7 +39,7 @@ paths:
       parameters:
       - in: "body"
         name: "srcUser"
-        required: false
+        required: true
```
Comment (liubao68) [inline swagger/swagger-generator/generator-springmvc/src/test/resources/schemas/mixupAnnotations.yaml]: Why change this test case? Can not figure out the related code change.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java (hunk 1 above)
Comment (acsukesh) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java]: its already added. private static void testRequiredBody(RestTemplate template, String microserviceName)   , server code @ AnnotationsTest.java,
TestMgr.check("No user data found",
+        template.postForObject(prefix + "/annotations/saysomething?prefix={prefix}",
+            null,

File: demo/demo-schema/src/main/resources/microservices/pojo/server.yaml (hunk 3 above)
Comment (acsukesh) [reply demo/demo-schema/src/main/resources/microservices/pojo/server.yaml]: here the test case is to check null when no body , but if we configure required as true it will always throw exception in client itself as null body not allowed as per swagger. so changed this to false as that is what is expected out of this test case.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java (hunk 1 above)
Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java]: I think there are still something missed. Let us view the consumer as a browser or postman.

Assume we have a service:
   void post(Person user)
Before this feature:
post man body: 
1. empty: provider will report invalidate parameter
2. null: (string null): provider will get a null Person instance

After this change:
1. empty: provider will get a null Person
2. null: provider will get a null Person(I am not tested)

And here you add if(arg !=null) statement, 
Before:
consumer send body: null string

Now: consumer send: empty 

This is a change, maybe you need to a manual test using postman. 

For null objects, consumer send empty is better than null, I think, so I can accept this change. But there are one consequence, that when required=true, users can send null string to provider but after change, user can never pass null to Person. 

Can you test and confirm? If my guess is correct, this code can not change, that's consumer always send null to provider for objects. required only work for 3rdparty clients.

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java (hunk 4 above)
Comment (liubao68) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java]: As mentioned in above latest comments, if here can pass "" to user parameter which indicate an empty body? null to indicate a null instance?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java (hunk 1 above)
Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java]: Maybe can pass "" as empty body using RestTemplate, but it's a bit strange send "" can not cast to Person.

Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java]: when required = true
1. Resttempalte or RPC always send empty body to provider and provider throws Exception
2. postman can send empty or null to provider, and provider throws Exception
when required = false
1. Resttempalte or RPC always send empty body to provider and provider get null
2. postman can send empty or null to provider, and provider get null

After change this is the expected value. I think is acceptable. When users not specify required = true, we can default to false.

**Comment by liubao68** [comment]
After some considerations about following scenario:
1. generated schema change: before required = false, and now required = true. 
          For this: users can increment microservice version and solve it
2. consume encoding null change: consumer sends null as a parameter
         For this: provider must upgrade SDK
It's very hard to make compatible solutions to this, and the compatible solutions will produce more things to hard to understand. As this two scenario have clear solution and rare, I accept this PR change and notice this change in RN.

**Comment by liubao68** [comment]
I created a new issue to track JAX-RS implementations
https://issues.apache.org/jira/browse/SCB-709
